storage-node/src/storage_manager.rs

```diff
@@ -13,24 +16,142 @@ use crate::{
 /// We should allow concurrent requests fed into the storage manager,
 /// which should be responsible for handling multiple requests at the
 /// same time.
+
 pub struct StorageManager<C: ParpulseCache> {
-    // TODO: Consider making the cache lock-free. See the comments for
-    // `ParpulseCache`.
-    cache: Arc<RwLock<C>>,
+    // TODO: Consider making the cache lock-free.
+    cache: RwLock<C>,
+    disk_manager: Mutex<DiskManager>,
+    // TODO: cache_base_path should be from config
+    // Cache_key = S3_PATH; Cache_value = (CACHE_BASE_PATH + S3_PATH, size)
+    cache_base_path: String,
 }
 
 impl<C: ParpulseCache> StorageManager<C> {
-    pub fn new(cache: C) -> Self {
+    pub fn new(cache: C, disk_manager: DiskManager, cache_base_path: String) -> Self {
         Self {
-            cache: Arc::new(RwLock::new(cache)),
+            cache: RwLock::new(cache),
+            disk_manager: Mutex::new(disk_manager),
+            cache_base_path,
         }
     }
 
-    pub async fn get_data(&self, _request: RequestParams) -> StorageResult<()> {
-        // TODO:
+    pub async fn get_data(&self, request: RequestParams) -> StorageResult<usize> {
         // 1. Try to get data from the cache first.
         // 2. If cache miss, then go to storage reader to fetch the data from
         // the underlying storage.
-        todo!()
+        // 3. If needed, update the cache with the data fetched from the storage reader.
+
+        // TODO: Support more request types.
+        let key = request.as_file().unwrap();
+        let mut cache = self.cache.write().await;
+
+        match cache.get(key) {
+            Some(value) => {
+                // Directly get data from the cache.
+                let cache_value_size = value.size();
+                let cache_file_path = value.as_value();
+
+                let mut output_iterator = self
+                    .disk_manager
+                    .lock()
```

Comment:
Actually I am not sure whether we should release this lock, since the purpose here to grab lock is to guarantee that the read data from iterator is counted to statistics without data race, but it seems release the lock immediately.
Also, do we need to record the read statistics? Because even if we only want to read the disk, we should grab this "write" lock to lock all the steps while reading...So currently I change the `self` of all the read functions in disk manager to immutable... (The statistics for read seem not very important : ( Bustub don't track their read statistics... Not sure ) But a better solution should be a statistics manager maybe!
Since we need to refactor, so I don't think we need to change the code. But do you think we should add some comments for furture hint here...?

Reply:
Just did minimal refactor based on the current code of the disk manager, where some of its methods (i.e. write_xxx) are still `&mut self`. If we remove the mutex here, we have to make it `fn get_data(&mut self)` in the storage manager. I don't think propagating the `&mut self` to such upper layer is a good idea since it limits the call to the storage manager itself. We can add a TODO.

Reply:
Yeah make sense!